documentation/projects/proposals/trust_and_safety/content_report_moderation/20240122-implementation_plan_moderation_metrics.md

```diff
@@ -0,0 +1,485 @@
+).values_list('decision_time', flat=True)
+average_decision_time = sum(time_differences, timedelta()) / len(time_differences) if time_differences else timedelta()
+
+p99_decision_time = reports_in_range.annotate(
+  decision_time=F('decision__created_at') - F('created_at')
+).aggregate(
+  percentile_99=Percentile('decision_time', 0.99)
+)['percentile_99']
+```
+
+#### Lists
+
+Lists are calculations that result in a list of items. For example, the most
+reported "leaderboards" are of this type.
+
+The following code can be used to calculate these leaderboards.
+
+```python
+from datetime import datetime, timedelta
+from django.utils import timezone
+from api.models import Report, Decision, Media
+
+# setup
+# =====
+# Assume n is an integer number of days set by the user.
+n_days_ago = timezone.now() - timedelta(days=n)
+reports_in_range = Report.objects.filter(created_at__gte=n_days_ago)
+
+# leaderboards
+# ============
+most_reported_media = reports_in_range.values('media_id').annotate(report_count=Count('id')).order_by('-report_count')[:10]
+most_reported_creator = reports_in_range.values('media__creator').annotate(report_count=Count('id')).order_by('-report_count')[:10]
+most_reported_source = reports_in_range.values('media__source').annotate(report_count=Count('id')).order_by('-report_count')[:10]
+```
+
+### Surface in Django Admin
+
+#### Values
+
+The admin site supports both adding
+[custom views](https://docs.djangoproject.com/en/5.0/ref/contrib/admin/#adding-views-to-admin-sites)
+and
+[overriding/extending existing templates](https://docs.djangoproject.com/en/5.0/ref/contrib/admin/#overriding-admin-templates).
+These approaches can be coupled together for a good experience in displaying
+these metrics to the moderators.
+
+Values related to image reports can be provided by editing the template that
+Django admin uses to render the "ImageReport" list view. Similar values for
+audio can be shown on the "AudioReport" view.
+
+#### Lists
+
+For the list of the most reported creator and most reported source, we can
+create pseudo-models with `managed = False`, register the model in the Django
+admin site and finally create custom filters to operate on the in-memory list.
+
+```python
+class Creator(models.Model):
+    # fields to uniquely identify a creator
+    name = models.CharField(max_length=100)
+    provider = models.CharField(max_length=100)
+
+    class Meta:
+        managed = False
+
+class CreatorAdmin(admin.ModelAdmin):
+    list_display = ('name', 'provider', 'report_count')
+    list_filter = ('provider',)
+
+    def report_count(self, obj):
+        return Report.objects.filter(media_provider=obj.provider).filter(media__creator=obj.name).count()
+
+    def get_queryset(self, request):
+        qs = super().get_queryset(request)
+        # Convert creators from media models into `Creator` models
+        return Media.objects.values('creator', 'provider').map(lambda x: Creator(name=x['creator'], provider=x['provider'])
```

Comment:
Not to be performing a code review here, but it looks like `qs` is unused? And this wouldn't iterate over _every_ Media record, would it? That seems like it might be a bad idea 😰

Reply:
This does iterate over every media item and is very inefficient, thanks for spotting that. I'll update the code example to make it more efficient by iterating over the reports (potentially even just the open ones).

Reply:
Updated the sample code in 26e46af.